### DISCUSSION TITLE
Счёт по префильтру

### DISCUSSION THREAD

**Discussion by Makhmut2445** [totumonline/totum-ru-issues-and-discussions#125] [Как это сделать?] [answered]
![image](https://user-images.githubusercontent.com/103171127/165751195-00970c7f-5877-4a8c-88c1-576791807da7.png)
Касаемо префильтра.
1 и 2 когда ставим префильтр, вычисляется и 4 и 5.. 

![image](https://user-images.githubusercontent.com/103171127/165751570-e9829be8-3b51-493b-8721-5b02d8b447ff.png)
Сделал так чтоб смотрела на 1 и 2 префильтры.. 

Нужно чтоб не зависимо от того какой префильтр поставлен, считала кол-во строк во всех полях, а не только то, в котором используется префильтр. Как можно сделать это по легче, чем использовать кучу if? Если префильтров 10, то код будет очень большим...

**Comment by ttm-alexeypolunin** [comment, marked as answer]
Изучение этого вопроса показало, что если у вас там 10 условий, то лучше использовать каскад из `if`.

Я попробовал создать через `while` и получилось вот так:

```
= : while(condition: $$iterator[cond] != true; action: $null; postaction: $end; iterator: "while"; limit: $limit)

zero: var(name: "exit"; value: "")
iterator: str`"row" + $#while`
null: ""
limit: 4
end: if(condition: $#while = math`$limit - 1`; condition: $$iterator[cond] = false; then: ""; else: $$row_exit)
		row_exit: $$iterator[exit]

row0: rowCreate(field: "cond" = false; field: "exit" = "null")
row1: rowCreate(field: "cond" = cond`#imya_polzovatelya = "Маша" && #kompaniya = 1`; field: "exit" = "exit_1")
row2: rowCreate(field: "cond" = cond`#imya_polzovatelya = "Петя" && #kompaniya = 1`; field: "exit" = "exit_2")
row3: rowCreate(field: "cond" = cond`#imya_polzovatelya = "Михаил" && #kompaniya = 3`; field: "exit" = "exit_3")

exit_1: "110"
exit_2: "120"
exit_3: "130"
```

Но это реально нужно если у вас 50 комбинаций условий.

**Немного упростил. Еще раз напоминаю, что вся эта дичь применима если у вас тьма условий — если 6-8 то лучше каскад из if.**

Итак по шагам, что здесь происходит.

1. `while` в коде вернет значение последнего `action` или `postaction`, поэтому в `$end` мы будем вычислять этот результат;
2. Нам нужно ограничить количество итераций по количеству проверяемых условий +1 и мы делаем это в `$limit`;
3. Почему `+1`, потому что что бы исполнился `postaction` должен быть хотя бы один выполненный `action` и это будет нулевой дополнительный `action`;
4. `action` нам нужен чисто технически, поэтому мы выполняем его в `$null`;
5. На каждом шаге мы в `condition` проверяем условие и проходим дальше если условие не выполняется. Те мы прерываем цикл и возвращаем ответ в момент, когда наткнемся на условие равное `true`;
6. `row1`, `row2` Я в пути! содержат условия и строку выхода;
7. Условия содержаться в [cond](https://ru.docs.totum.online/syntax#sx-cond);
8. `$$` в цикле это мы берем значение строки определенное другой строкой, в этом случае строка `$iterator` определяет из какой строки нам брать условие на текущем шаге в зависимости от `iterator`;
9. `iterator` на каждом шаге перещелкивается на единицу, начинается от нуля;
10. В `$end` у нас есть дополнительное условие связанное с тем, что `postaction` выполнится в том случае если ни один из `condition` не выполнился (в нашем случае там условие инвертировано и получается они все выполнились);
11. Те для последнего итератора мы повторно проверяем его условие и если оно `false` то `пустота`;
12. В остальных случаях результат выходной строки.

**Reply by Makhmut2445** [reply]
Хорошо было бы. Ждём пояснения

**Reply by ttm-alexeypolunin** [reply]
Чуть-чуть упростил и дополнил описание.
